src/python-common/ceph/deployment/service_spec.py

```diff
@@ -471,6 +471,9 @@ def validate(self):
         if not self.service_type:
             raise ServiceSpecValidationError('Cannot add Service: type required')
 
+        if self.service_type in ['mds', 'rgw', 'nfs', 'iscsi'] and not self.service_id:
+            raise ServiceSpecValidationError('Cannot add Service: id required')
+
```

Comment:
Is there a chance that there are existing specs in the store that violate this validation?

Reply:
nice, this is what I meant to do as well.

> Is there a chance that there are existing specs in the store that violate this validation?

They probably is. Is the `validate` method being run on already existing specs though? If not, this would only raise for new or exported(and re-imported) specs.

Reply:
> nice, this is what I meant to do as well.
> 
> > Is there a chance that there are existing specs in the store that violate this validation?
> 
> They probably is. Is the `validate` method being run on already existing specs though?

right: `__init__` doesn't call `validate()`:

https://github.com/ceph/ceph/blob/3a757a4d670ed531c34537889744116a9959ae1c/src/python-common/ceph/deployment/service_spec.py#L399-L411

and https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/inventory.py doesn't call `.validate()` either. 

should we add it? If yes, how do we handle specs that fail the validation?

> If not, this would only raise for new or exported(and re-imported) specs.

Reply:
> should we add it? If yes, how do we handle specs that fail the validation?

we probably only should do that if we have a clear path for communicating these kind of issues. some way of indicating that an already applied spec is not valid anymore..

otoh, we probably shouldn't invalidate existing specs to begin with.. maybe adding an internal tracking id for already existing specs isn't the worst move.

Reply:
it actually _is_ called on existing specs here:

https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/schedule.py#L67

Reply:
which is not as bad as in `load()` . I think #35456 plus some HEALTH_WARN integration would be enough here.

Reply:
> Is there a chance that there are existing specs in the store that violate this validation?

If so, they would have likely broke the `orch ls` etc commands...

> which is not as bad as in `load()` . I think #35456 plus some HEALTH_WARN integration would be enough here.

This actually makes a lot of sense in the general case where the cluster state could change and we might want to (re)validate before another placement...

Reply:
so I think we're good, as long as we provide a way for users to remove that service after they did an upgrade.